doc/man/man3/Player.3
.TH "Player" 3 "Version 1.0" "Zaxxon" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Player \- The player character, also contains plane controls\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fR#include <Player\&.h>\fP
.PP
Inherits \fBCharacter\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBPlayer\fP (sf::Texture *, unsigned int)"
.br
.RI "Create player character at position and load resources for them\&. "
.ti -1c
.RI "void \fBupdate\fP (sf::RenderWindow &, int, float)"
.br
.RI "Run all the logic and controls for player\&. "
.ti -1c
.RI "void \fBkill\fP () override"
.br
.RI "Modify variables for when player dies or respawns\&. "
.ti -1c
.RI "void \fBresetPos\fP (int zOffset=0)"
.br
.RI "Reset the players position to default\&. "
.ti -1c
.RI "void \fBdrawHitmarker\fP ()"
.br
.ti -1c
.RI "bool \fBisAlive\fP ()"
.br
.RI "Check if player is alive\&. "
.ti -1c
.RI "bool \fBisMissileable\fP ()"
.br
.RI "Returns whether the horizontal missile should shoot at the player or not\&. "
.ti -1c
.RI "void \fBrestartMissileTimer\fP ()"
.br
.in -1c

Public Member Functions inherited from \fBCharacter\fP
.in +1c
.ti -1c
.RI "\fBCharacter\fP (sf::Texture *)"
.br
.RI "Create a new character who will have sprite from spritesheet\&. "
.ti -1c
.RI "\fB~Character\fP () override"
.br
.RI "Clean up memory related to the \fBCharacter\fP class\&. "
.ti -1c
.RI "void \fBupdate\fP (sf::RenderWindow &, float) override"
.br
.RI "Move the character by velocity and redraw them to the screen\&. "
.ti -1c
.RI "std::vector< \fBCharacterBullet\fP * > & \fBgetBullets\fP ()"
.br
.RI "Get the currently fired bullets from this character\&. "
.ti -1c
.RI "unsigned int \fBgetSizeIndex\fP () const"
.br
.RI "Get public size index of this character\&. "
.ti -1c
.RI "void \fBkillBullet\fP (int)"
.br
.RI "Get rid of bullet by index\&. "
.ti -1c
.RI "void \fBsetPos\fP (sf::Vector3f)"
.br
.RI "Set position for the character itself\&. "
.in -1c

Public Member Functions inherited from \fBEntity\fP
.in +1c
.ti -1c
.RI "\fBEntity\fP ()"
.br
.RI "Setup \fBEntity\fP class variables\&. "
.ti -1c
.RI "virtual \fB~Entity\fP ()=0"
.br
.RI "Clean up memory related to \fBEntity\fP class\&. "
.ti -1c
.RI "virtual void \fBkill\fP (\fBAnimation::Anim\fP \fBanimation\fP=\fBAnimation::CHARACTER_DEATH\fP)"
.br
.RI "Run death animation code for the killed entity\&. "
.ti -1c
.RI "int \fBgetAnimationState\fP ()"
.br
.ti -1c
.RI "void \fBsetPos\fP (sf::Vector3f pos)"
.br
.ti -1c
.RI "sf::Vector3f \fBgetPos\fP ()"
.br
.ti -1c
.RI "sf::FloatRect \fBgetBounds\fP ()"
.br
.ti -1c
.RI "sf::Vector2f \fBgetSpritePos\fP ()"
.br
.RI "Get the position of the related sprite for this entity\&. "
.in -1c
.SS "Additional Inherited Members"


Protected Member Functions inherited from \fBCharacter\fP
.in +1c
.ti -1c
.RI "void \fB_getSizeIndex\fP (unsigned int &)"
.br
.RI "Get the height sector that character is in\&. Underscore infront indicates that it is not public\&. 0-3 3 indicates smallest, 0 indicates biggest\&. "
.ti -1c
.RI "sf::Vector3f \fBgetVelocity\fP () const"
.br
.RI "Get the velocity of character\&. "
.ti -1c
.RI "void \fBsetVelocity\fP (sf::Vector3f)"
.br
.RI "Change the velocity of this character\&. "
.ti -1c
.RI "void \fBsetBullet\fP (sf::IntRect)"
.br
.RI "Change how the bullets from this character look\&. "
.ti -1c
.RI "void \fBupdateBullets\fP (sf::RenderWindow &window, float)"
.br
.RI "Calls \fBCharacterBullet::update()\fP on each bullet & deletes bullets that are not in the window\&. "
.in -1c

Protected Attributes inherited from \fBCharacter\fP
.in +1c
.ti -1c
.RI "const int \fByMax\fP = 140"
.br
.ti -1c
.RI "const int \fByMin\fP = 69"
.br
.ti -1c
.RI "const int \fBxMin\fP = \-160"
.br
.ti -1c
.RI "const int \fBxMax\fP = 0"
.br
.ti -1c
.RI "sf::Vector3f \fBvelocity\fP"
.br
.ti -1c
.RI "std::vector< \fBCharacterBullet\fP * > \fBbullets\fP"
.br
.ti -1c
.RI "unsigned int \fBsizeIndex\fP"
.br
.in -1c

Protected Attributes inherited from \fBEntity\fP
.in +1c
.ti -1c
.RI "sf::Texture * \fBspriteSheet\fP"
.br
.ti -1c
.RI "sf::Sprite * \fBsprite\fP"
.br
.ti -1c
.RI "\fBAnimation\fP \fBanimations\fP"
.br
.ti -1c
.RI "bool \fBdead\fP = false"
.br
.ti -1c
.RI "sf::Text \fBdebugText\fP"
.br
.ti -1c
.RI "sf::Font \fBdebugFont\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
The player character, also contains plane controls\&. 
.PP
Definition at line \fB10\fP of file \fBPlayer\&.h\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Player::Player (sf::Texture * texture, unsigned int startPos)"

.PP
Create player character at position and load resources for them\&. 
.PP
\fBParameters\fP
.RS 4
\fItexture\fP 
.br
\fIstartPos\fP 
.RE
.PP

.PP
Definition at line \fB10\fP of file \fBPlayer\&.cpp\fP\&.
.SH "Member Function Documentation"
.PP 
.SS "void Player::drawHitmarker ()\fR [inline]\fP"

.PP
Definition at line \fB18\fP of file \fBPlayer\&.h\fP\&.
.SS "bool Player::isAlive ()"

.PP
Check if player is alive\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.PP
Definition at line \fB186\fP of file \fBPlayer\&.cpp\fP\&.
.SS "bool Player::isMissileable ()"

.PP
Returns whether the horizontal missile should shoot at the player or not\&. 
.PP
\fBReturns\fP
.RS 4

.RE
.PP

.PP
Definition at line \fB176\fP of file \fBPlayer\&.cpp\fP\&.
.SS "void Player::kill ()\fR [override]\fP, \fR [virtual]\fP"

.PP
Modify variables for when player dies or respawns\&. 
.PP
Implements \fBCharacter\fP\&.
.PP
Definition at line \fB156\fP of file \fBPlayer\&.cpp\fP\&.
.SS "void Player::resetPos (int zOffset = \fR0\fP)"

.PP
Reset the players position to default\&. 
.PP
\fBParameters\fP
.RS 4
\fIzOffset\fP 
.RE
.PP

.PP
Definition at line \fB166\fP of file \fBPlayer\&.cpp\fP\&.
.SS "void Player::restartMissileTimer ()\fR [inline]\fP"

.PP
Definition at line \fB23\fP of file \fBPlayer\&.h\fP\&.
.SS "void Player::update (sf::RenderWindow & window, int stage, float gameSpeed)"

.PP
Run all the logic and controls for player\&. 
.PP
\fBParameters\fP
.RS 4
\fIwindow\fP 
.br
\fIinSpace\fP 
.br
\fIgameSpeed\fP 
.RE
.PP

.PP
Definition at line \fB42\fP of file \fBPlayer\&.cpp\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Zaxxon from the source code\&.


src/Entity/Character/Player/Player.cpp
#include "Player.h"
#include "Background/Background.h"


/// <summary>
/// Create player character at position and load resources for them.
/// </summary>
/// <param name="texture"></param>
/// <param name="startPos"></param>
Player::Player(sf::Texture* texture, unsigned int startPos) : Character(texture)
{
	sf::IntRect a = sf::IntRect(8, 13, 23, 23);
	for (unsigned int i = 0; i < 3; i++)
	{
		for (unsigned int j = 0; j < 4; j++)
		{
			playerTextures[i][j] = a;
			a.left += 23;
		}
	}

	this->sprite->setTextureRect(playerTextures[0][0]);
	this->setPos(sf::Vector3f(0, 69, static_cast<int>(startPos) * -1.33333f));
	this->shadow.setTexture(*texture);
	this->shadow.setTextureRect(sf::IntRect(352, 18, 22, 13));
	this->shadow.setColor(sf::Color::Black);
	this->hitmarker.setTexture(*texture);
	this->hitmarker.setTextureRect(sf::IntRect(320, 84, 12, 11));

	// Prepare bullet sound
	bulletBuffer.loadFromFile("res/sfx/08.wav");
	bulletSound.setBuffer(bulletBuffer);
}


/// <summary>
/// Run all the logic and controls for player.
/// </summary>
/// <param name="window"></param>
/// <param name="inSpace"></param>
/// <param name="gameSpeed"></param>
void Player::update(sf::RenderWindow& window, int stage, float gameSpeed)
{
	if (!alive)
		return;

	// If alive after this point
	// Update texture
	unsigned int planeVertical = 0;
	_getSizeIndex(sizeIndex);

	// Keys
	sf::Vector3f tempVelocity = getVelocity();
	constexpr float acceleration = 1.f / 5.f;

	if (leftPressed() && getPos().x < xMax && tempVelocity.x < 1.f)
		tempVelocity.x += acceleration;
	else if (rightPressed() && getPos().x > xMin && tempVelocity.x > -1.f)
		tempVelocity.x -= acceleration;
	else if (tempVelocity.x < 0)
		tempVelocity.x += acceleration;
	else if (tempVelocity.x > 0)
		tempVelocity.x -= acceleration;

	if (tempVelocity.x < acceleration && tempVelocity.x > -acceleration)
		tempVelocity.x = 0;

	if (upPressed() && tempVelocity.y < 0.6f)
		tempVelocity.y += acceleration / 0.5f;
	else if (downPressed() && tempVelocity.y > -0.6f)
		tempVelocity.y -= acceleration / 0.5f;
	else if (tempVelocity.y < 0)
		tempVelocity.y += acceleration / 0.5f;
	else if (tempVelocity.y > 0)
		tempVelocity.y -= acceleration / 0.5f;

	// Set plane sprite var before we reset velocity for bounds
	if (tempVelocity.y == 0)
		planeVertical = 0;
	else if (tempVelocity.y < 0)
		planeVertical = 1;
	else if (tempVelocity.y > 0)
		planeVertical = 2;

	if (getPos().y <= 69.6 && tempVelocity.y < 0)
		tempVelocity.y = 0;
	else if (getPos().y >= 139.4 && tempVelocity.y > 0)
		tempVelocity.y = 0;

	if (stage != 1)
		sizeIndex = 0;
	sprite->setTextureRect(playerTextures[planeVertical][sizeIndex]);

	// Spawn bullets
	if (zPressed() && bulletCD.getElapsedTime().asMilliseconds() >
		BULLET_COOLDOWN / gameSpeed)
	{
		bulletCD.restart();

		bullets.
			push_back(new CharacterBullet(spriteSheet, getPos(), sizeIndex));

		bulletSound.play();
	}
#ifndef NDEBUG
	if (sf::Keyboard::isKeyPressed(sf::Keyboard::K))
		std::cout << getPos().x << " " << getPos().y << " " << getPos().z <<
			"\n";
#endif

	if (stage != 3)
	{
		if (stage != 1)
			tempVelocity.z = -1.3f; //for moving with background
		else // space moves 2/3 speed
			tempVelocity.z = -1.3f * 0.66f;
	}
	else
		tempVelocity.z = 0;

	// Position updates
	setVelocity(tempVelocity);
	shadow.setPosition(
		translateTo2d(sf::Vector3f(getPos().x - 5, 2 * 224 / 3, getPos().z)));
	hitmarker.setPosition(sprite->getPosition() + sf::Vector2f(40, -15));

	// Shadow before character
	if (stage != 1)
		window.draw(shadow);
	else if (hitmarkerTimer.getElapsedTime().asMilliseconds() < 150)
		window.draw(hitmarker);

	// if you stay above 80 for 3s and stage is initial then you get missile after you
	if (getPos().y > 80 || stage != 0)
		missileTimer.restart();

	// Drawing
	Character::update(window, gameSpeed);
	// updating position using velocity, draw character
	updateBullets(window, gameSpeed);

#ifndef NDEBUG
	debugText.setString(std::to_string(static_cast<int>(getPos().x)) + " " +
		std::to_string(static_cast<int>(getPos().y)) + " " + std::to_string(
			static_cast<int>(getPos().z))
	);
	debugText.setPosition(sprite->getPosition());
	window.draw(debugText);
#endif // !NDEBUG
}


/// <summary>
/// Modify variables for when player dies or respawns.
/// </summary>
void Player::kill()
{
	alive = !alive;
}


/// <summary>
/// Reset the players position to default.
/// </summary>
/// <param name="zOffset"></param>
void Player::resetPos(int zOffset)
{
	setPos(sf::Vector3f(getPos().x, getPos().y, zOffset * -1.33333f));
}


/// <summary>
/// Returns whether the horizontal missile should shoot at the player or not.
/// </summary>
/// <returns></returns>
bool Player::isMissileable()
{
	return missileTimer.getElapsedTime().asSeconds() > 5 && getPos().z < -300;
}


/// <summary>
/// Check if player is alive.
/// </summary>
/// <returns></returns>
bool Player::isAlive()
{
	return alive;
}
